sys/hoon.hoon

```diff
@@ -11598,23 +11397,3309 @@
+    ^-  ximage
+    ::
+    |^  (main [0 ~ ~] top-type)
+    ::
+    ++  main
+      |=  [st=xtable ty=type]
+      ^-  [xkey xtable]
+      ::
+      =/  old  (~(get by type-map.st) ty)             ::  already done
+      ?^  old  [u.old st]
+      ::
+      =^  res=xkey  st  (post-xray st ty ~)
+      ::
+      :-  res
+      ?-  ty
+        %void      (set-xray-xdat st res %void)
+        %noun      (set-xray-xdat st res %noun)
+        [%atom *]  (set-xray-xdat st res ty)
+        [%cell *]  =^  hed=xkey  st  (main st p.ty)
+                   =^  tyl=xkey  st  (main st q.ty)
+                   (set-xray-xdat st res [%cell hed tyl])
+        [%core *]  =^  d=xdat   st  (xray-core [p.ty q.ty] st)
+                   (set-xray-xdat st res d)
+        [%face *]  =^  i=xkey  st  (main st q.ty)
+                   (set-xray-xdat st res [%face p.ty i])
+        [%fork *]   =^  br  st  ((traverse-set xtable type xkey) [st p.ty] main)
+                   (set-xray-xdat st res [%fork br])
+        [%hint *]  =^  ref      st  (main st q.ty)
+                   =^  updated  st  (hint st p.ty (focus-on st res))
+                   (set-xray-xdat (replace-xray st updated) res [%pntr ref])
+        [%hold *]  =^  ref  st  (main st ~(repo ut ty))
+                   (set-xray-xdat st res [%pntr ref])
+      ==
+    ::
+    ::  Analyze a %hint type.
+    ::
+    ::  This updates the `helps`, `studs`, and/or `recipe` fields of the
+    ::  given xray.
+    ::
+    ++  hint
+      |=  [st=xtable [subject-of-note=type =note] x=xray]
+      ^-  [xray xtable]
+      ?-  -.note
+        %help  :_  st  x(helps (~(put in helps.x) p.note))
+        %know  :_  st  x(studs (~(put in studs.x) p.note))
+        %made  =^  recipe  st
+                 ?~  q.note  [[%direct p.note] st]
+                 =^  params=(list xkey)  st
+                   |-  ^-  [(list xkey) xtable]
+                   ?~  u.q.note  [~ st]
+                   =/  tsld  [%tsld [%limb %$] [%wing i.u.q.note]]
+                   =/  part  (~(play ut subject-of-note) tsld)
+                   =^  this  st  (main st part)
+                   =^  more  st  $(u.q.note t.u.q.note)
+                   [[this more] st]
+                 [[%synthetic p.note params] st]
+               :_  st  x(recipes (~(put in recipes.x) recipe))
+      ==
+    ::
+    ::  Analyze a core.
+    ::
+    ::  When we analyze the context, we decrement `core-depth`. If that
+    ::  ever hits zero, we substitute `%noun` for the type of the context.
+    ::
+    ::  The reason that we switch the varience to %gold is because the
+    ::  core we're creating isn't an actual core, we're just using the arms
+    ::  of this core as a namespace in which to evaluate each arm.
+    ::
+    ::  Also, in general, there's no way to determine the type of an arm
+    ::  of a wet core, so we just assign all wet arms the type `%noun`.
+    ::
+    ::  This seems to work in practice, but I don't think it's actually
```

Comment:
Do you remember why this is unsound? I don't understand the danger

Reply:
This is worth thinking about more, but there's no right answers, and this works for now.

It *is* wrong, and it's worth understanding about how far-reaching (or not) the effect of that is.